lib/gc/Transforms/GPU/GpuUtils.h

```diff
@@ -0,0 +1,241 @@
+    T first[] = {*begin, std::accumulate(begin + 2, end, *(begin + 1),
+                                         std::multiplies<>())};
+    adjustTiles(total, first, first + 2, minSize);
+    adjustTiles(total / *first, begin + 1, end, minSize);
+    *begin = *first;
+    return;
+  }
+
+  --end;
+  T a = *begin;
+  T b = *end;
+  bool swap = a < b;
+  if (swap) {
+    std::swap(a, b);
+  }
+
+  a = ceilPow2(a);
+  b = floorPow2(b);
+
+  if (a * b <= total) {
+    *begin = swap ? b : a;
+    *end = swap ? a : b;
+    return;
+  }
+
+  double ratio = static_cast<double>(a) / static_cast<double>(b);
+  T x = static_cast<T>(std::sqrt(total)) * static_cast<T>(std::sqrt(ratio));
+  x = std::min(ceilPow2(x), std::min(a, total));
+  T y = std::min(floorPow2(total / x), b);
+
+  // Adjust x and y to get the closest ratio
+  if (auto diff =
+          std::abs(ratio - static_cast<double>(x) / static_cast<double>(y));
+      y >= 2 && x * 2 <= a &&
+      std::abs(ratio - static_cast<double>(x * 2) /
+                           static_cast<double>(y / 2)) < diff) {
+    x *= 2;
+    y /= 2;
+  } else if (x >= 2 && y * 2 <= b &&
+             std::abs(ratio - static_cast<double>(x / 2) /
+                                  static_cast<double>(y * 2)) < diff) {
+    x /= 2;
+    y *= 2;
+  }
+
+  if (y < minSize && a >= minSize && b >= minSize) {
+    if (auto newX = ceilPow2(total / minSize); newX >= minSize) {
+      x = std::min(newX, a);
+      y = minSize;
+    }
+  }
+
+  *begin = swap ? y : x;
+  *end = swap ? x : y;
+}
+
+template <typename T, unsigned N>
+static void adjustTiles(T totalSize, SmallVector<T, N> &tiles,
+                        T minSize = static_cast<T>(1)) {
+  adjustTiles(totalSize, tiles.begin(), tiles.end(), minSize);
+}
+
+// Check recursively if the specified operation has an operand that
+// depends on a result of a previous operation, matching the predicate.
+template <unsigned MaxDepth = std::numeric_limits<unsigned>::max()>
+bool isOperandDependsOnOp(bool (*predicate)(Operation *), Operation *operation,
+                          unsigned depth = 0) {
+  for (auto operand : operation->getOperands()) {
+    if (auto op = operand.getDefiningOp();
+        op &&
+        (predicate(op) || (depth < MaxDepth &&
+                           isOperandDependsOnOp(predicate, op, depth + 1)))) {
+      return true;
+    }
+  }
+  return false;
+}
+
+// Check recursively if there are any operation, matching the predicate, that
+// depends on the result of the specified operation.
+template <unsigned MaxDepth = std::numeric_limits<unsigned>::max()>
+bool isOpDependsOnResult(bool (*predicate)(Operation *), Operation *operation,
+                         unsigned depth = 0) {
+  for (auto res : operation->getResults()) {
+    for (auto u : res.getUsers()) {
+      if (predicate(u) ||
+          (depth < MaxDepth && isOpDependsOnResult(predicate, u, depth + 1))) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
```

Comment:
This looks like it can get pretty expensive. Can it be replaced by some standard data flow analysis?